### PR TITLE
Limit number of analyze for one query

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* Break on analyze stuck on complex query

Ref https://github.com/ClickHouse/ClickHouse/issues/21557

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Interpreters/Context.h [hunk 1]
```diff
@@ -364,6 +364,15 @@ class Context: public std::enable_shared_from_this<Context>
     // Top-level OpenTelemetry trace context for the query. Makes sense only for a query context.
     OpenTelemetryTraceContext query_trace_context;
 
+    /// Some counters for current query execution.
+    /// Most of them is workarounds and should be removed in the future.
```
Comment (alexey-milovidov) [inline src/Interpreters/Context.h:368]:
```suggestion
    /// Most of them are workarounds and should be removed in the future.
```

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 2]
```diff
@@ -447,6 +448,15 @@ InterpreterSelectQuery::InterpreterSelectQuery(
 
     auto analyze = [&] (bool try_move_to_prewhere)
     {
+        if (context->hasQueryContext())
+        {
+            size_t & current_query_analyze_count = context->getQueryContext()->kitchen_sink.analyze_counter;
+            current_query_analyze_count++;
```
Comment (alexey-milovidov) [inline src/Interpreters/InterpreterSelectQuery.cpp:454]:
```suggestion
            ++current_query_analyze_count;
```

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 3]
```diff
@@ -447,6 +448,15 @@ InterpreterSelectQuery::InterpreterSelectQuery(
...
+            if (current_query_analyze_count >= settings.max_pipeline_depth)
+                throw DB::Exception(ErrorCodes::TOO_DEEP_PIPELINE, "Query analyze overflow. Try to increase `max_pipeline_depth` or simplify the query" );
```
Comment (alexey-milovidov) [inline src/Interpreters/InterpreterSelectQuery.cpp:456]: Also, we planned to use this counter to inhibit the second analysis step.

**Review by alexey-milovidov — APPROVED** [review]

File: src/Interpreters/InterpreterSelectQuery.cpp (hunk 3 above)
Comment (vdimir) [reply src/Interpreters/InterpreterSelectQuery.cpp:456]: In which cases second analysis step should be inhibited?

Comment (alexey-milovidov) [reply src/Interpreters/InterpreterSelectQuery.cpp:456]: It was said that the second analysis step is only an optimization. (Although I'm not sure)
If we already have made too many analyses, we can avoid further optimizations.

Also you can remove exponential blowup by memorizing the ASTs that were already optimized.

Comment (vdimir) [reply src/Interpreters/InterpreterSelectQuery.cpp]: Changed it to use `max_pipeline_depth / 10` as a the threshold to disable the second analyze and `max_pipeline_depth` to cancel the query at all.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@kssenii
RabbitMQ tests have failed: https://s3.amazonaws.com/clickhouse-test-reports/38185/1e974b55ea63430a059f70f06892865b4caafd7e/integration_tests__release__[2/2].html

@Felixoid
any hypothesis about the DNS issue and how to avoid it?
https://s3.amazonaws.com/clickhouse-test-reports/38185/1e974b55ea63430a059f70f06892865b4caafd7e/performance_comparison_[3/4]/runlog.log

**Comment by tavplubix** [comment]
@alexey-milovidov, RabbitMQ tests failure was actually related to changes in this PR:
https://play.clickhouse.com/play?user=play#c2VsZWN0IAp0b1N0YXJ0T2ZEYXkoY2hlY2tfc3RhcnRfdGltZSkgYXMgZCwKY291bnQoKSwgZ3JvdXBVbmlxQXJyYXkocHVsbF9yZXF1ZXN0X251bWJlciksIGFueShyZXBvcnRfdXJsKQpmcm9tIGNoZWNrcyB3aGVyZSAnMjAyMi0wNi0wMScgPD0gY2hlY2tfc3RhcnRfdGltZSBhbmQgdGVzdF9uYW1lIGxpa2UgJyV0ZXN0X3N0b3JhZ2VfcmFiYml0bXElJyBhbmQgdGVzdF9zdGF0dXMgaW4gKCdGQUlMJykgZ3JvdXAgYnkgZCBvcmRlciBieSBkIGRlc2M=

**Comment by vdimir** [comment]
@tavplubix thanks for noticing, I'll check

**Comment by vdimir** [comment]
StorageRabbitMQ uses a global context for storage, and that's why the counter overflowed. Probably it's better to find another solution than the counter in context, but for now, I can't come up with it.

https://github.com/ClickHouse/ClickHouse/blob/1c0d2677673a82336641605d08b743964eb3dadc/src/Storages/RabbitMQ/StorageRabbitMQ.cpp#L1056-L1058

**Comment by alexey-milovidov** [comment]
@vdimir please resubmit this PR.

**Comment by vdimir** [comment]
@alexey-milovidov 
Here it is https://github.com/ClickHouse/ClickHouse/pull/40334, the same code, but with disabling for StorageRabbitMQ, because it reuses the same query context.

**Comment by kssenii** [comment]
> but with disabling for StorageRabbitMQ, because it reuses the same query context

for writing to materialised view? may be just create a new query context there each time?

**Pull request merged**